core/storage/couchStorage.go

```diff
@@ -0,0 +1,691 @@
+	//store.checkObjects()
+}
+
+// Cleanup erase the on disk Bolt database only for ESS and test
+func (store *CouchStorage) Cleanup(isTest bool) common.SyncServiceError {
+	return nil
+}
+
+// GetObjectsToActivate returns inactive objects that are ready to be activated
+func (store *CouchStorage) GetObjectsToActivate() ([]common.MetaData, common.SyncServiceError) {
+	return nil, nil
+}
+
+// StoreObject stores an object
+// If the object already exists, return the changes in its destinations list (for CSS) - return the list of deleted destinations
+func (store *CouchStorage) StoreObject(metaData common.MetaData, data []byte, status string) ([]common.StoreDestinationStatus, common.SyncServiceError) {
+	id := getObjectCollectionID(metaData)
+	var dests []common.StoreDestinationStatus
+	var deletedDests []common.StoreDestinationStatus
+
+	if metaData.DestinationPolicy != nil {
+		metaData.DestinationPolicy.Timestamp = time.Now().UTC().UnixNano()
+	}
+
+	if status == common.NotReadyToSend || status == common.ReadyToSend {
+		// The object was receieved from a service, i.e. this node is the origin of the object:
+		// set its instance id and create destinations array
+		newID := store.getInstanceID()
+		metaData.InstanceID = newID
+		if data != nil && !metaData.NoData && !metaData.MetaOnly {
+			metaData.DataID = newID
+		}
+
+		var err error
+		dests, deletedDests, err = createDestinationsFromMeta(store, metaData)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	existingObject := &couchObject{}
+	if err := store.getOne(id, existingObject); err != nil {
+		if !strings.HasPrefix(err.Error(), "Not Found") {
+			return nil, &Error{fmt.Sprintf("Failed to retrieve object's status. Error: %s.", err)}
+		}
+		existingObject = nil
+	}
+
+	if existingObject != nil {
+		if (metaData.DestinationPolicy != nil && existingObject.MetaData.DestinationPolicy == nil) ||
+			(metaData.DestinationPolicy == nil && existingObject.MetaData.DestinationPolicy != nil) {
+			return nil, &common.InvalidRequest{Message: "Can't update the existence of Destination Policy"}
+		}
+
+		if metaData.MetaOnly {
+			metaData.DataID = existingObject.MetaData.DataID
+			metaData.ObjectSize = existingObject.MetaData.ObjectSize
+			metaData.ChunkSize = existingObject.MetaData.ChunkSize
+		}
+		if metaData.DestinationPolicy != nil {
+			dests = existingObject.Destinations
+		}
+
+	}
+
+	newObject := couchObject{ID: id, MetaData: metaData, Status: status, PolicyReceived: false,
+		RemainingConsumers: metaData.ExpectedConsumers,
+		RemainingReceivers: metaData.ExpectedConsumers, Destinations: dests}
+
+	if existingObject != nil {
+		if err := store.upsert(newObject, true); err != nil {
+			return nil, &Error{fmt.Sprintf("Failed to store an object. Error: %s.", err)}
+		}
+	} else {
+		if err := store.upsert(newObject, false); err != nil {
+			return nil, &Error{fmt.Sprintf("Failed to store an object. Error: %s.", err)}
+		}
+	}
+
+	if !metaData.NoData && data != nil {
```

Comment:
Is there no way in couch to add an object and its attachment at the same time, in one database transaction? If not, it means that a caller could see the object but not the attachment, though this wont happen very often. Perhaps Mongo has this same problem. This is something to keep in mind, and something to include in the design document.

Reply:
According to CouchDB documentation, it is possible to add an object and its attachment at the same time. But when I tried to do this with Kivik PUT (in a sample program), the attachment wasn't getting stored with the object. It was not throwing any error, but when I checked the document in the database, it did not show an attachment. I also tried getting this attachment using GetAttachment() and I got a 'Document is missing attachment' error. I tried this multiple times and got the same result so this could possibly be a Kivik bug or currently not supported...
I am only able to add an attachment using PutAttachment() function for an already existing object.

I will include this in the design doc, this would be the kind of 'eventual consistency' specific errors we had discussed before, right? Mongo may have the same problem but it would probably not be seen because of its strict consistency policy

Reply:
I suspect Mongo would still have this problem even with strict consistency because the issue has more to do with the ACID properties of the database. Even with strict consistency, 2 calls to the database will each occur within their own transaction. In order to avoid the visibility problem, the caller (ess) would have to provide both the object and the data on the same call so that the DB could make both updates in the same ACID transaction. If this isnt possible then there will always be a small window where the object is visible but not the data/attachment.